### ISSUE TITLE
💡 Roadmap: The Road to v1.0

### ISSUE THREAD

**Issue by FredKSchott** [withastro/astro#1222]
## Background

There was a good discussion in Discord (#ideas-and-suggestions channel) about how to prioritize different projects on the roadmap. The @snowpackjs/maintainers-core team has had a few discussions internally about this, but we haven't really set it out in an explicit list yet. So, here's my attempt to capture that list.

The goal is to come up with a document that can be saved to ROADMAP.md and referenced in Discord, issues, and Open Collective (having a well defined Roadmap is essential for larger corporate sponsorships). 

This issue (everything below, starting with the "Roadmap" header) is the draft of that document. Once accepted, I'll commit the issue into the repo.

Please leave comments & suggest changes before we merge!

---

# Roadmap 

**Current Mission: The Road to v1.0**  
Our current mission is to launch an Astro v1.0 release later this year. To do that, there are certain features that must be shipped before we can consider Astro "feature complete" and worthy of the v1.0.0 version release. This will also give us a chance to gather user feedback on early features, make changes faster, and eventually finalize behavior for that v1.0.0 release that will not be easily changed going forward.

## Goals

- Stability
- Developer experience
- (as we near the end) Performance

## Tasks

- [x] In browser REPL (owner: @natemoo-re #1359)
- [x] VSCode extension  (owner: @matthewp)
- [x] `astro preview` command (owner: @FredKSchott #1226)
- [x] an npm story for sharable components (owner: @FredKSchott, @jonathantneal)
- [x] *To finalize:* build performance (owner: not yet claimed)
- [x] *To finalize:* build output security, XSS, etc. (owner: not yet claimed) (#1828, #1827)
- [ ] *To finalize:* build optimization behavior (bundling, critical CSS) (owner: not yet claimed)
- [ ] *To finalize:* console output across build and dev (owner: not yet claimed)
- [ ] *To finalize:* any breaking default behavior (ex: style hoisting) (owner: not yet claimed)
- [ ] *To finalize:* all syntax, directive naming, etc. (owner: not yet claimed)
- [ ] Documentation: fully translated docs site into Tier 1 languages (owner: not yet claimed)
- [ ] Documentation: guides and examples for all popular CMS (owner: not yet claimed)
- [ ] prettier plugin (owner: not claimed, @jasikpark has done work in the past)

#### Handled by new compiler + vite
- [x] New WASM compiler (owner: @natemoo-re, @drwpow #1406)
- [x] components in Markdown (owner: @natemoo-re #1406)
- [x] `<script use:variables={{...}}>` (owner: @natemoo-re #1406)
- [x] finalize style & script behavior in new compiler (owner: @natemoo-re #1406)
- [ ] exports from Astro components (owner: @natemoo-re #1142 #1406)

#### Bonus: Nice to have, but not required
- [ ] `<Image>` component (owner: not yet claimed)
- [ ] fetch respects HTTP cache control headers (owner: @jasikpark)
- [ ] Memoization helpers (owner: not yet claimed)
- [ ] *TBD:* `.html` components & page support (owner: not yet claimed, #1227)
- [ ] build & dev server customization, plugin system (owner: not yet claimed)

_Note: This list of tasks is non-binding: some things might be learned as we go that influence the document. Some things may need to go through the RFC process before getting added to the roadmap. We will try to keep this roadmap up to date to the best of our ability (see something out of date? create a PR!)._

_Note: Got a feature suggestion that you don't see on this list? You should still feel welcome to post in Discord (channel: #ideas-and-suggestions) or create an RFC. The core maintainer team may not be able to prioritize things outside of this list, but that doesn't mean that we can't discuss it, schedule it for a future roadmap, or support you in adding it yourself!_

**Comment by matthewp** [comment]
XSS is in in two bullet points, I think one's a dupe.

**Comment by RafidMuhymin** [comment]
How much importance should be given to paving the way for building a better ecosystem?

**Comment by FredKSchott** [comment]
> How much importance should be given to paving the way for building a better ecosystem?

Not sure what you mean, can you give some examples?

**Comment by RafidMuhymin** [comment]
> > How much importance should be given to paving the way for building a better ecosystem?
> 
> Not sure what you mean, can you give some examples?

Astro is still too young and it has a very very small ecosystem. So, how much importance should be given to enlarging the ecosystem? For example, build APIs, plugins, themes, starters, etc.

**Comment by RafidMuhymin** [comment]
I am willing to claim the `<Image>` component. I will use the `imagetools-core` package to achieve this instead of reinventing the wheel. I have already started working on this in my personal project.

**Comment by jasikpark** [comment]
I would say that *is* the goal of this roadmap in a way - we want to work out the kinks for 1.0 so that then we can provide a stable basis for community development to work off of!

**Comment by RafidMuhymin** [comment]
> I would say that _is_ the goal of this roadmap in a way - we want to work out the kinks for 1.0 so that then we can provide a stable basis for community development to work off of!

Reasonable.

**Comment by FredKSchott** [comment]
> I am willing to claim the `<Image>` component. I will use the `imagetools-core` package to achieve this instead of reinventing the wheel. I have already started working on this in my personal project.

Awesome! This could probably live entirely in a "user-land" project to start, so that you are free to experiment in your own `src/components/Image.astro`  file + project without worrying about learning the Astro codebase. If you can get something good working in your project that you're happy with, we could look into integrating into Astro!

There's an RFC that outlines some of the features we were interested in: #492

**Comment by FredKSchott** [comment]
> I would say that _is_ the goal of this roadmap in a way - we want to work out the kinks for 1.0 so that then we can provide a stable basis for community development to work off of!

Yup, the biggest item here for that is "make sure the npm publish story is solid". The rest of the roadmap is about getting to a stable foundation in v1.0, so that anyone feels comfortable building on top of it.

**Comment by RafidMuhymin** [comment]
> Awesome! This could probably live entirely in a "user-land" project to start, so that you are free to experiment in your own `src/components/Image.astro` file + project without worrying about learning the Astro codebase. If you can get something good working in your project that you're happy with, we could look into integrating into Astro!
> 
> There's an RFC that outlines some of the features we were interested in: #492

Thank you for your helpful suggestions! I will post all the updates in the discord channel.

**Comment by itskitto** [comment]
Could we/I get a round off of what we would consider the popular CMSs? As I transition into a new schedule, this may be something I can tackle now

**Comment by FredKSchott** [comment]
Want to start a thread in #ideas-and-suggestions in Discord? I could see that getting noisy and want to keep this issue focused on 1.0 readiness.

**Comment by mihilmy** [comment]
How can I contribute to this roadmap? I think this is a very promising project and excited to see it going to v1.

**Comment by FredKSchott** [comment]
I'm sure @jasikpark could use some help with the Prettier plugin. It's one of the more requested features that we get!

Other than that, most of the "nice to haves" don't have an owner yet and could be great to investigate.

Finally, anything in the RFC "accepted" column is fair game and has already passed through an RFC approval process.
https://github.com/snowpackjs/astro/projects/3

**Comment by AsyncBanana** [comment]
I could help with the CMS examples, although I do not know if I could do all of it myself, depending on how many CMS examples you are thinking of. Also, would they be example projects paired with a summary to get started, or would each have a whole page on the docs along with the example project?

**Comment by FredKSchott** [comment]
@AsyncBanana That would be great! Mind pinging @natemoo-re in Discord, maybe in the #dev or #start-contributing channels? I think he's interested in kicking off some integration work to document different CMS, tools, etc.

**Comment by matthewp** [comment]
I've identified blockers for the 1.0 release and documented them here: https://github.com/snowpackjs/astro/issues/1855

**Comment by FredKSchott** [comment]
Update: we've kicked off a dedicated "v1.0 API" discussion in our new RFC repo: https://github.com/withastro/rfcs/discussions/1

**Comment by Siilwyn** [comment]
What about minifying HTML?

**Comment by matthewp** [comment]
Should we close this issue given that the RFC discussion exists now?

**Comment by FredKSchott** [comment]
+1, I just got confused by this the other day. All discussion is now in https://github.com/withastro/rfcs/discussions/1
